### ISSUE TITLE
[R] MKL-DNN support:  "Unknown exception" in mx.nd.internal.as.array

### ISSUE THREAD

**Issue by Crunchy9** [apache/incubator-mxnet#15420]
## Description
After successful build with MKL-DNN R package, when I start traning  exception occurse.

I'm using R version 3.6.0 (2019-04-26)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 8.1 x64 (build 9600)

## Build info (Required if built from source)

Compiler: Visual Studio 15 (14.16)
MKL 2019.4.245
mklml_win_2019.0.5.20190502.zip

MXNet commit hash: 51acd4d328c6da475c058806565d93bcddee7402

## Error Message:

> (...) use 4 threads for decoding..
> Start training with 1 devices
> Error in mx.nd.internal.as.array(nd) : Unknown exception
> Calls: mx.model.FeedForward.create ... as.array.MXNDArray -> mx.nd.internal.as.array -> .External
> Execution halted

## Minimum reproducible example
Choose any network to train.

## What have you tried to solve it?
Don't know what to do with this exception... :-/

**Comment by vdantu** [comment]
@mxnet-label-bot add [Windows, R, Training]

**Comment by vdantu** [comment]
@anirudhacharya : Would you be able to help with this?

**Comment by pengzhao-intel** [comment]
@Crunchy9 could you provide a reproducible case?

**Comment by Crunchy9** [comment]
This will be hard, but... `traceback()` 

> 9: stop(list(message = "Unknown exception", call = mx.nd.internal.as.array(nd), 
       cppstack = list(file = "", line = -1L, stack = "C++ stack not available on this system")))
8: .External(list(name = "InternalFunction_invoke", address = <pointer: 0x0000000006998270>, 
       dll = list(name = "Rcpp", path = ".../R/win-library/3.6/Rcpp/libs/x64/Rcpp.dll", 
           dynamicLookup = TRUE, handle = <pointer: 0x000000006abc0000>, 
           info = <pointer: 0x0000000006961060>), numParameters = -1L), 
       <pointer: 0x000000000892b990>, ...)
7: mx.nd.internal.as.array(nd)
6: as.array.MXNDArray(res)
5: as.array(res)
4: feval(label, pred)
3: metric$update(label = labels[[i]], pred = preds[[i]], state = train.metric)
2: mx.model.train(symbol, ctx, input.shape, output.shape, params$arg.params, 
       params$aux.params, begin.round, num.round, optimizer = optimizer, 
       train.data = X, eval.data = eval.data, metric = eval.metric, 
       epoch.end.callback = epoch.end.callback, batch.end.callback = batch.end.callback, 
       kvstore = kvstore, fixed.param = fixed.param, verbose = verbose, 
       metric_cpu = metric_cpu)
1: mx.model.FeedForward.create(softmax, initializer = mx.init.Xavier(factor_type = "in", 
       magnitude = 2), X = dane, ctx = devices, num.round = 300, 
       begin.round = epoch + 1, eval.data = NULL, optimizer = mx.opt.create("sgd", 
           learning.rate = 0.005, momentum = 0.9, wd = 0, lr_scheduler = fs), 
       eval.metric = mx.metric.accuracy, batch.end.callback = mx.callback.log.speedometer(...)), epoch.end.callback = mx.callback.save.checkpoint(paste0("...", 
           ...), 1))
Surprisingly, examples [tutorial](https://mxnet.incubator.apache.org/versions/master/tutorials/r/fiveMinutesNeuralNetwork.html) works fine.

**Comment by TaoLv** [comment]
@Crunchy9 We notice that the document for Windows R setup says:

> Note: packages for 3.6.x are not yet available. Install 3.5.x of R from CRAN.

http://mxnet.incubator.apache.org/versions/master/install/windows_setup.html#install-the-mxnet-package-for-r

Have you ever tried a 3.5.x version of R?

**Comment by Crunchy9** [comment]
Yes, and it's strange. Installed side by side R 3.6.0 build from source and downloaded AWS for 3.5.3.

> Error in mx.varg.io.ImageRecordIter(list(...)) : 
  [07:47:26] c:\incubator-mxnet\src\io\iter_image_recordio_2.cc:254: ImageRec need opencv to process

**Comment by TaoLv** [comment]
Seems it's another error now. I just checked the code. Do you have `USE_OPENCV=1` in you make command line when build the package with MKL-DNN?

**Comment by Crunchy9** [comment]
Of course. But why it fails with  official release now?
`c:\build_mxnet\with_mkldnn\incubator-mxnet\build>cmake -G "Visual Studio 15 Win6
4" .. -DUSE_CUDA=0 -DUSE_CUDNN=0 -DUSE_NVRTC=0 -DUSE_OPENCV=1 -DUSE_OPENMP=1 -DU
SE_PROFILER=1 -DUSE_BLAS=mkl -DUSE_LAPACK=1 -DUSE_DIST_KVSTORE=0 -DCUDA_ARCH_NAM
E=All -DUSE_MKLDNN=1 -DCMAKE_BUILD_TYPE=Release -DMKL_ROOT="C:\Program Files (x8
6)\IntelSWTools\compilers_and_libraries\windows\mkl"`
I've checked with` InstanceBatchNorm` and `mx.symbol.BatchNorm`, nothing changes.

**Comment by TaoLv** [comment]
I'm sorry. What do you mean by "I've checked with InstanceBatchNorm and mx.symbol.BatchNorm, nothing changes."?  Do you still have the issue described in the title of this issue?

**Comment by Crunchy9** [comment]
Yes, error is the same despite of batch norm type.
For R 3.6.0 `install.packages("https://s3.ca-central-1.amazonaws.com/jeremiedb/share/mxnet/CPU/3.6/mxnet.zip", repos = NULL)` I used this version with no OpenCV errors.

**Comment by TaoLv** [comment]
Hi @jeremiedb, seems you contributed the package. Is it supposed to work in this case?

**Comment by Crunchy9** [comment]
Reproducible case (sort of):
`data <- mx.symbol.Variable("data")`
`conv <- mx.symbol.Convolution(data = data, kernel = c(7, 7) , stride = c(2, 2), pad = c(0, 0), num.filter = 24, name = paste0("test", "_conv1"))`
`act  <- mx.symbol.LeakyReLU(data = conv, act.type = "elu", name = paste0("test", "_act1"))
`
`softmax <- FC(data = act, name = "out")`

Error:
> Start training with 1 devices
Error in mx.nd.internal.dispatch.Ops(.Generic, e1, e2) : 
  [10:40:29] c:\build_mxnet\with_mkldnn\incubator-mxnet\src\operator\tensor\../elemwise_op_common.h:135: Check failed: assign(&dattr, vec.at(i)): Incompatible attr in node  at 1-th input: expected [8], got [16]
Calls: mx.model.FeedForward.create ... Ops.MXNDArray -> mx.nd.internal.dispatch.Ops -> .External
Execution halted

Output from R CMD looks normal.

> ** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
>* MD5 sums
zip I/O error: No such file or directory
zip error: Temporary file failure (D:/zih8Nowg)
running 'zip' failed
>* DONE (mxnet)

But interaction with MKL-DNN looks wierd.

**Comment by TaoLv** [comment]
I'm totally lost. Seems the log is different from what you shared in the issue title and description. Can you articulate what's the problem now and please share a complete reproducer rather than a code snippet.

**Comment by Crunchy9** [comment]
I'm lost too, because every time something new is popping out. Seems to me it's a OpenCV problem (?).

1.  Build R package with MKL-DNN, OpenCV 4.1.0, Visual Studio 15.
2.  Reproducible example (at last), data from [Kaggle](https://www.kaggle.com/jessicali9530/caltech256).
`python im2rec.py --resize 224 --encoding .jpg --quality 60`
`library(mxnet)
data<-mx.io.ImageRecordIter(path.imgrec="D:/256_ObjectCategories/caltech256_train.rec", path.imglist="D:/256_ObjectCategories/caltech256_train.lst",  batch.size = 8, data_shape = c(224, 224, 3))
data <- mx.symbol.Variable("data")
conv <- mx.symbol.Convolution(data, kernel = c(7, 7) , stride = c(2, 2), pad = c(0, 0), num.filter = 24, name = paste0("test", "_conv1"))
act  <- mx.symbol.LeakyReLU(conv, act.type = "leaky", name = paste0("test", "_act1"))
fc <- mx.symbol.FullyConnected(act, num_hidden = 256, name = paste0("test", "_FC"))
softmax <- mx.symbol.SoftmaxOutput(fc, name = "softmax")
devices <- mx.cpu()
model <- mx.model.FeedForward.create(softmax,
										initializer=mx.init.Xavier(factor_type = "in", magnitude=2),
										X=data, ctx=devices, num.round=2, begin.round=epoch+1,
										eval.data=NULL,
										optimizer=mx.opt.create("sgd", learning.rate = 0.005,  momentum = 0.9, wd = 0, lr_scheduler = NULL),
										eval.metric=mx.metric.accuracy)`

Now I got...

> Error in mx.model.init.iter(X, y, batch.size = array.batch.size, is.train = TRUE) : 
  Need to provide parameter y for training with R arrays.

**Comment by RuilinZhuIntel** [comment]
@Crunchy9 Hi, please try this code for training on your kaggle dataset,
library(mxnet)
dataset<-mx.io.ImageRecordIter(path.imgrec="C:\\Users\\ruilinzh\\Downloads\\caltech-256-image-dataset\\data_train.rec", path.imglist="C:\\Users\\ruilinzh\\Downloads\\caltech-256-image-dataset\\data_train.lst", batch.size = 8, data_shape = c(224, 224, 3))
data <- mx.symbol.Variable("data") 
conv <- mx.symbol.Convolution(data, kernel = c(7, 7) , stride = c(2, 2), pad = c(0, 0), num.filter = 24, name = paste0("test", "_conv1")) 
act <- mx.symbol.LeakyReLU(conv, act.type = "leaky", name = paste0("test", "_act1")) 
fc <- mx.symbol.FullyConnected(act, num_hidden = 256, name = paste0("test", "_FC")) 
softmax <- mx.symbol.SoftmaxOutput(fc, name = "softmax") 
devices <- mx.cpu() 
model <- mx.model.FeedForward.create(softmax, initializer=mx.init.Xavier(factor_type = "in", magnitude=2), X=dataset, ctx=devices, num.round=2, begin.round=epoch+1, eval.data=NULL, optimizer=mx.opt.create("sgd", learning.rate = 0.005, momentum = 0.9, wd = 0, lr_scheduler = NULL), eval.metric=mx.metric.accuracy)

**Comment by Crunchy9** [comment]
@RuilinZhuIntel, @TaoLv Thanks guys!
